amaranth/hdl/ast.py

```diff
@@ -110,7 +145,7 @@ def cast(obj, *, src_loc_at=0):
         while True:
             if isinstance(obj, Shape):
                 return obj
-            elif isinstance(obj, ShapeCastable):
+            elif ShapeCastable in obj.__class__.__mro__:
```

Comment:
This is the only part of the PR/fix that gives me pause--this part is kind of ... Extremely obscure. Is there any better way to make this check? I was hoping that overriding `__instancecheck__` but not `__subclasscheck__` would be a viable solution, but having thought about it more that seems even worse than this.

I can see having to just accept this but it doesn't have to make me happy :s

Reply:
Re: the `type(obj).mro()` suggestion (per @mwkmwkmwk)—this is almost OK, but unfortunately still not:

```python
>>> from amaranth.lib.enum import Enum as AmaranthEnum
>>> class EnumA(AmaranthEnum): X = 0
... 
>>> EnumA.__class__.__mro__
(<class 'amaranth.lib.enum.EnumMeta'>, <class 'amaranth.hdl.ast.ShapeCastable'>, <class 'enum.EnumType'>, <class 'type'>, <class 'object'>)
>>> type(EnumA).mro()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unbound method type.mro() needs an argument
>>>
```

`mro()` is defined on classes, not metaclasses: i.e. it's an instance method on `type`. For all call-sites, where the object we're checking may have a metaclass as its `type()`—which includes passing any enum class—this will fail. All call-sites may receive enum classes.

(It's worth noting this behaviour isn't due to any interaction with `ABCMeta`.)

We _can_ use `type(obj).__mro__`, which contains 50% fewer underscores than what the PR currently proposes, so that's a start. (I've also found one site where we can actually use `isinstance` without causing recursive bad.)

One 'solution' is to make a helper and be done with it. The naming is awkward: `ShapeCastable.actual_instance(o)`? `_utils.in_type_mro(o, c)`? `c in _utils.tmro(o)`?

Another is to reduce the number of places we have to do this to less than 3, and then it's under threshold for refactoring!

More thoughts below.